### PR TITLE
240 review nncell and nncombinatorial models + Fixed SAN bugs

### PR DESCRIPTION
**Pull request by gbg141**
This pull request addresses the following points:

- Refactorization of cell and combinatorial models, decoupling the readout function from their implementation (analogous to what was already implemented for simplicial and hypergraph architectures, PRs #250 and #249). Models, notebooks and tests have been updated.
- Solved two bugs in SAN implementation --it was not possible to train it on GPU.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -121,3 +121,8 @@ checks = [
     "EX01",
     "SA01"
 ]
+exclude = [
+    '\.undocumented_method$',
+    '\.__repr__$',
+    '\.__init__$',
+]
```
Comment (ffl096) [inline pyproject.toml:128]: Please revert this, `__init__` and `__repr__` should be documented.

File: topomodelx/nn/cell/can.py [hunk 2]
```diff
@@ -47,14 +47,14 @@ def __init__(
         in_channels_0,
         in_channels_1,
         out_channels,
-        num_classes,
         dropout=0.5,
-        heads=3,
+        heads=2,
```
Comment (ffl096) [inline topomodelx/nn/cell/can.py:51]: The documentation for `heads` in the docstring needs to be updated as well.

File: topomodelx/nn/cell/can.py [hunk 3]
```diff
@@ -129,8 +127,8 @@ def forward(
 
         Returns
         -------
-        torch.Tensor
-            Output prediction for the cell complex.
+        x_1: torch.Tensor, shape = (num_pooled_edges, heads * out_channels)
```
Comment (ffl096) [inline topomodelx/nn/cell/can.py:130]: Singular return values do not have a name.

File: topomodelx/nn/cell/cwn.py [hunk 4]
```diff
@@ -107,23 +104,4 @@ def forward(
...
-        zero_dimensional_cells_mean = torch.nanmean(x_0, dim=0)
-        zero_dimensional_cells_mean[torch.isnan(zero_dimensional_cells_mean)] = 0
-
-        # Return the sum of the averages
-        return (
-            two_dimensional_cells_mean
-            + one_dimensional_cells_mean
-            + zero_dimensional_cells_mean
-        )
+        return (x_0, x_1, x_2)
```
Comment (ffl096) [inline topomodelx/nn/cell/cwn.py:107]: Parentheses are unnecessary here.

File: topomodelx/nn/combinatorial/hmc.py [hunk 5]
```diff
@@ -72,7 +69,7 @@ def forward(
         neighborhood_2_to_2,
         neighborhood_0_to_1,
         neighborhood_1_to_2,
-    ) -> torch.Tensor:
+    ):
```
Comment (ffl096) [inline topomodelx/nn/combinatorial/hmc.py:72]: Please don't just remove the return type but change it appropriately.

File: topomodelx/nn/combinatorial/hmc.py [hunk 6]
```diff
@@ -18,30 +18,31 @@ class HMC(torch.nn.Module):
...
 
     def __init__(
         self,
         channels_per_layer,
-        num_classes,
         negative_slope=0.2,
         update_func_attention="relu",
         update_func_aggregation="relu",
-    ) -> None:
+    ):
```
Comment (ffl096) [inline topomodelx/nn/combinatorial/hmc.py:36]: Revert please.

**Review by ffl096 — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 7]
```diff
@@ -18,6 +18,7 @@ __pycache__/
 docs/build/
 data/
 venv_*/
+.DS*
```
[the remark this answers is not in the document]
Comment (gbg141) [reply .gitignore]: Done.

File: pyproject.toml (hunk 1 above)
Comment (gbg141) [reply pyproject.toml]: I have reverted it, but currently `__init__` methods are not documented in TopoModelX. This results in failing the tests for any class that you modify.

File: topomodelx/nn/cell/can.py (hunk 2 above)
Comment (gbg141) [reply topomodelx/nn/cell/can.py:51]: Done, thank you.

File: topomodelx/nn/cell/can.py (hunk 3 above)
Comment (gbg141) [reply topomodelx/nn/cell/can.py]: Done.

File: topomodelx/nn/combinatorial/hmc.py (hunk 5 above)
Comment (gbg141) [reply topomodelx/nn/combinatorial/hmc.py]: Adapted.

File: topomodelx/nn/cell/cwn.py (hunk 4 above)
Comment (gbg141) [reply topomodelx/nn/cell/cwn.py]: Done.

**Review by ffl096 — APPROVED** [review]

**Pull request merged**